llsd/base.py

```diff
@@ -399,53 +395,71 @@ def _getc(self, num=1):
 
     # map char following escape char to corresponding character
     _escaped = {
-        b'a': b'\a',
-        b'b': b'\b',
-        b'f': b'\f',
-        b'n': b'\n',
-        b'r': b'\r',
-        b't': b'\t',
-        b'v': b'\v',
+        ord(b'a'): ord(b'\a'),
+        ord(b'b'): ord(b'\b'),
+        ord(b'f'): ord(b'\f'),
+        ord(b'n'): ord(b'\n'),
+        ord(b'r'): ord(b'\r'),
+        ord(b't'): ord(b'\t'),
+        ord(b'v'): ord(b'\v'),
     }
 
     def _parse_string_delim(self, delim):
         "Parse a delimited string."
-        parts = bytearray()
-        found_escape = False
-        found_hex = False
-        found_digit = False
-        byte = 0
+        insert_idx = 0
+        delim_ord = ord(delim)
+        # Preallocate a working buffer for the decoded string output
+        # to avoid allocs in the hot loop.
+        decode_buff = self._decode_buff
+        # Cache these in locals, otherwise we have to perform a lookup on
+        # `self` in the hot loop.
+        buff = self._buffer
+        read_idx = self._index
+        cc = 0
         while True:
-            cc = self._getc()
-            if found_escape:
-                if found_hex:
-                    if found_digit:
-                        found_escape = False
-                        found_hex = False
-                        found_digit = False
-                        byte <<= 4
-                        byte |= _hex_as_nybble(cc)
-                        parts.append(byte)
-                        byte = 0
+            try:
+                cc = buff[read_idx]
+                read_idx += 1
+
+                if cc == _BACKSLASH_ORD:
+                    # Backslash, figure out if this is an \xNN hex escape or
+                    # something like \t
+                    cc = buff[read_idx]
+                    read_idx += 1
+                    if cc == _X_ORD:
+                        # It's a hex escape. char is the value of the two
+                        # following hex nybbles
+                        cc = int(chr(buff[read_idx]), 16) << 4
+                        read_idx += 1
+                        cc |= int(chr(buff[read_idx]), 16)
+                        read_idx += 1
```

Comment:
This `\xNN` parsing could be further optimized, but how much of that do we expect?

Reply:
Depends, I believe the viewer hex escapes any bytes with the high bit set, so it'd be a problem when dealing anything outside ascii range. Were you thinking something like `cc = int(str(buff[read_idx:read_idx+1]), 16)`?

Reply:
Handled by https://github.com/secondlife/python-llsd/pull/1/commits/1bb6812e117d80387fcdffbb1403cf872e6513bb, which ends up being a lot faster (and also faster than two separate reads with concat and one `int()`) by a lot. It's 30% faster than both `main` and the previous commit with the synthetic testcase:

```python
llsd.parse(b'[' + b','.join([b'"' + b''.join([b"\\xc3\\xa9"] * 10) + b'"'] * 100000) + b']')
```